influxdb/metric.go

```diff
@@ -0,0 +1,224 @@
+	// by applying an offset. We'll need to apply filtering once cumulative
+	// metrics come into play.
+	return m.metricDescriptor(), m.convertTimeSeries(points), nil
+}
+
+// buildQuery creates a InfluxDB query from the metric query definition,
+// wrapped in the given time-interval, inclusive of start time.
+func (m *Metric) buildQuery(startTime, endTime time.Time) client.Query {
+	query := fmt.Sprintf(
+		"SELECT * FROM (%s) WHERE time >= %s AND time < %s",
+		m.config.Query,
+		strconv.FormatInt(startTime.UnixNano(), 10),
+		strconv.FormatInt(endTime.UnixNano(), 10))
+
+	return client.NewQuery(query, m.config.Database, "ns")
+}
+
+func (m *Metric) metricDescriptor() *metricpb.MetricDescriptor {
+	return &metricpb.MetricDescriptor{
+		Type:        m.StackdriverName(),
+		MetricKind:  metricpb.MetricDescriptor_GAUGE,
+		ValueType:   metricpb.MetricDescriptor_DOUBLE,
+		Description: fmt.Sprintf("InfluxDB query: %s", m.Name),
+		DisplayName: m.config.Query,
+	}
+}
+
+func (m *Metric) convertTimeSeries(points []point) []*monitoringpb.TimeSeries {
+	ts := make([]*monitoringpb.TimeSeries, 0, len(points))
+	for _, p := range points {
+		ts = append(ts, &monitoringpb.TimeSeries{
+			Metric:     &metricpb.Metric{Type: m.StackdriverName()},
+			Resource:   &monitoredres.MonitoredResource{Type: "global"},
+			MetricKind: metricpb.MetricDescriptor_GAUGE,
+			ValueType:  metricpb.MetricDescriptor_DOUBLE,
+			Points:     []*monitoringpb.Point{p.convertPoint()},
+		})
+	}
+	return ts
+}
+
+type point struct {
+	timestamp time.Time
+	value     float64
+}
+
+// parseSeriePoints parses points from an InfluxDB series into a slice of
+// timestamp-value pairs.
+func parseSeriePoints(serie models.Row) ([]point, error) {
+	if len(serie.Columns) != 2 {
+		return nil, fmt.Errorf("serie has columns %s, expected only 2 columns", serie.Columns)
+	} else if serie.Columns[0] != "time" {
+		return nil, fmt.Errorf("serie has first column '%s', expected 'time'", serie.Columns[0])
+	}
+
+	var points []point
+	for _, p := range serie.Values {
+		t, ok := p[0].(json.Number)
+		if !ok {
+			return nil, fmt.Errorf("failed to cast %v to json.Number", p[0])
+		}
+		unixNano, err := t.Int64()
+		if err != nil {
+			return nil, err
+		}
+
+		v, ok := p[1].(json.Number)
+		if !ok {
+			return nil, fmt.Errorf("failed to cast %v to json.Number", p[1])
+		}
+		// Since the column types are not specified, we can only assume float64.
+		val, err := v.Float64()
+		if err != nil {
+			return nil, err
+		}
+
+		points = append(points, point{time.Unix(0, unixNano), val})
+	}
+
+	return points, nil
+}
+
+// convertPoint converts a parsed InfluxDB point into a Stackdriver point.
+func (p *point) convertPoint() *monitoringpb.Point {
+	// For gauge metrics without time aggregations, we can treat the timestamps
+	// given by Influx as EndTime for the Stackdriver point.
+	return &monitoringpb.Point{
+		Interval: &monitoringpb.TimeInterval{
+			EndTime: p.protoTimestamp(),
```

Comment:
I think `timestamp.TimestampProto()` is the recommended way to convert `time.Time` to a `Timestamp` proto.

Reply:
done